consensus/tendermint/rule_proposal_and_polka_current_test.go

```diff
@@ -0,0 +1,98 @@
+package tendermint
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestProposalAndPolkaCurrent(t *testing.T) {
+	t.Run("Line 36: lock and broadcast precommit when step is prevote", func(t *testing.T) {
+		stateMachine := setupStateMachine(t, 4, 3)
+		currentRound := newTestRound(t, stateMachine, 0, 0)
+
+		committedValue := value(42)
+
+		// Initialise the round
+		currentRound.start()
+
+		// Prevotes are collected
+		currentRound.validator(0).prevote(&committedValue)
+		currentRound.validator(1).prevote(&committedValue)
+
+		// Receives proposal, go straight to precommit
+		currentRound.validator(0).proposal(committedValue, -1).expectActions(
+			currentRound.action().scheduleTimeout(prevote),
+			currentRound.action().broadcastPrevote(&committedValue),
+			currentRound.action().broadcastPrecommit(&committedValue),
+		)
+
+		assert.True(t, stateMachine.state.lockedValueAndOrValidValueSet)
+		assertState(t, stateMachine, height(0), round(0), precommit)
+	})
+
+	t.Run("Line 36: record valid value even if we don't prevote it", func(t *testing.T) {
+		stateMachine := setupStateMachine(t, 4, 1)
+		currentRound := newTestRound(t, stateMachine, 0, 0)
+		nextRound := newTestRound(t, stateMachine, 0, 1)
+		committedValue := value(42)
+
+		// Initialise the round
+		currentRound.start()
+
+		// Prevotes are collected but not proposal
+		currentRound.processTimeout(propose).expectActions(
+			currentRound.action().broadcastPrevote(nil),
+		)
+		currentRound.validator(2).prevote(&committedValue)
+		currentRound.validator(3).prevote(&committedValue)
+
+		currentRound.processTimeout(prevote).expectActions(
+			currentRound.action().broadcastPrecommit(nil),
+		)
+
+		// Receives proposal and prevote from validator 0, record valid value and round
+		currentRound.validator(0).proposal(committedValue, -1)
+		currentRound.validator(0).prevote(&committedValue)
+		assertState(t, stateMachine, height(0), round(0), precommit)
```

Comment:
I think something that would be very helpful when reading tests, is adding a small comment indicating when and which "line" we enter

```suggestion
		currentRound.validator(3).prevote(&committedValue) // line34: schedule timeout

		currentRound.processTimeout(prevote).expectActions(
			currentRound.action().broadcastPrecommit(nil),
		)

		// Receives proposal and prevote from validator 0, record valid value and round
		currentRound.validator(0).proposal(committedValue, -1)
		currentRound.validator(0).prevote(&committedValue) // line 36
		assertState(t, stateMachine, height(0), round(0), precommit)
```

Reply:
Agree, for line 34 comment, I think I'll add an expectActions there, so it can be clearer.